### ISSUE TITLE
VAE does not generate reasonable images

### ISSUE THREAD

**Issue by FomalhautB** [lucidrains/DALLE-pytorch#45]
I am trying to train the VAE for the COCO dataset. But the generated output looks strange (input above, output below):

![image](https://user-images.githubusercontent.com/14837467/108221880-8a759100-7138-11eb-9c4c-9c3619b43e74.png)

My code:

```
import os
import json
import numpy as np
from PIL import Image
import matplotlib.pyplot as plt
import torch
from torch import optim
from torch.utils.data import Dataset, DataLoader
from dalle.dalle_pytorch import DiscreteVAE

IMAGE_SIZE = 256
TRAIN_ANNOT_PATH = '...'
VAL_ANNOT_PATH = '...'
TRAIN_IMAGE_PATH = '...'
VAL_IMAGE_PATH = '...'

class COCODataset(Dataset):
    def __init__(self, annot_path, image_path, image_size=256):
        self._image_path = image_path
        
        with open(annot_path) as file:
            json_file = json.load(file)
        
        self._image_size = image_size
        self._metadata = json_file['images']
        self._captions = {entry['image_id']: entry for entry in json_file['annotations']}
    
    def __getitem__(self, index):
        metadata = self._metadata[index]
        caption = self._captions[metadata['id']]
        image_path = os.path.join(self._image_path, metadata['file_name'])
        image = Image.open(image_path).convert('RGB')
        image = self._crop_image(image)
        x = np.asarray(image) / 255
        return torch.Tensor(x).permute(2, 0, 1), caption['caption']
    
    def _crop_image(self, image):
        width, height = image.size
        min_length = min(width, height)
        
        # center crop
        left = (width - min_length)/2
        top = (height - min_length)/2
        right = (width + min_length)/2
        bottom = (height + min_length)/2
        image = image.crop((left, top, right, bottom))
        
        # resize
        image = image.resize((self._image_size, self._image_size))
        
        return image
    
    def __len__(self):
        return len(self._metadata)
        

def tensor_to_image(tensor):
    tensor = tensor.permute(1, 2, 0)
    arr = ((tensor.numpy())*255).astype(np.uint8)
    return Image.fromarray(arr)

BATCH_SIZE = 16
train_dataset = COCODataset(TRAIN_ANNOT_PATH, TRAIN_IMAGE_PATH, image_size=IMAGE_SIZE)
val_dataset = COCODataset(VAL_ANNOT_PATH, VAL_IMAGE_PATH, image_size=IMAGE_SIZE)

train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, num_workers=4)
val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE, num_workers=4)

def train_vae(epoch, model, loader, optimizer, device):
    model.train()
    
    for index, (x, _) in enumerate(loader):
        x = x.to(device)
        optimizer.zero_grad()
        
        loss = model(x, return_loss=True)
        loss.backward()
        optimizer.step()
        
        print(f'Epoch {epoch:2} [{index:4}/{len(loader):4}]: Loss: {loss:.4}')

DEVICE = 'cuda'
vae = DiscreteVAE(
    image_size = IMAGE_SIZE,
    num_layers = 3,          # number of downsamples - ex. 256 / (2 ** 3) = (32 x 32 feature map)
    num_tokens = 8192,       # number of visual tokens. in the paper, they used 8192, but could be smaller for downsized projects
    codebook_dim = 512,      # codebook dimension
    hidden_dim = 64,         # hidden dimension
    num_resnet_blocks = 1,   # number of resnet blocks
    temperature = 0.9,       # gumbel softmax temperature, the lower this is, the harder the discretization
    straight_through = False # straight-through for gumbel softmax. unclear if it is better one way or the other
).to(DEVICE)

EPOCHS = 2
optimizer = optim.Adam(vae.parameters(), lr=0.001, weight_decay=0.0)

for epoch in range(1, EPOCHS+1):
    print(f'Epoch {epoch}/{EPOCHS}'.center(50, '='))
    train_vae(epoch, vae, train_loader, optimizer, DEVICE)

vae.eval()
x, y = val_dataset[0]
codes = vae.get_codebook_indices(x.unsqueeze(0).cuda())
generated = vae.decode(codes)[0].detach().cpu()
plt.imshow(tensor_to_image(x))
plt.show()
plt.imshow(tensor_to_image(generated))
plt.show()
```

I have trained the VAE for 2 epochs (10k+ steps). The training loss decreased from 3.0+ to about 0.07. The output looks like the monotone color even on training set images. Am I doing anything wrong?

**Comment by lucidrains** [comment]
@FomalhautB Hi Zai! That may be due to the recent extra loss term I added to the VAE. I've decreased the strength of it in the latest patch, but you can also play around with it (or turn it off to have it the way it was before, with a value of `0`) https://github.com/lucidrains/DALLE-pytorch/blob/main/dalle_pytorch/dalle_pytorch.py#L77

**Comment by FomalhautB** [comment]
> @FomalhautB Hi Zai! That may be due to the recent extra loss term I added to the VAE. I've decreased the strength of it in the latest patch, but you can also play around with it (or turn it off to have it the way it was before, with a value of `0`) https://github.com/lucidrains/DALLE-pytorch/blob/main/dalle_pytorch/dalle_pytorch.py#L77
> 
> You can control this with the keyword `kl_div_loss_weight` on VAE initialization

Thank you! After setting `kl_div_loss_weight=0`, it works properly!
![image](https://user-images.githubusercontent.com/14837467/108242622-66707a80-714d-11eb-9fb8-f753783636b2.png)

**Comment by lucidrains** [comment]
@FomalhautB ok! i was told to have this value be non-zero to prevent codebook collapse, but glad to hear it is working!

**Comment by zijin-gu** [comment]
I set kl_div_loss_weight=0 but got something like this
![reference](https://user-images.githubusercontent.com/55098043/150908559-cb391cb8-77e7-4cd7-bc19-205b29f72a88.png)
![generated](https://user-images.githubusercontent.com/55098043/150908573-3bb9f42b-6f2a-4f59-a76a-0c9d6e2a469e.png)
is there any possible reasons? 
And why we need to set the weight for KL divergence to 0?
